### PR TITLE
Improve default services start strategy.

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this PR do?
Improve services start strategy. Add detail message when user create compose environment with circle dependency to itself.

@garagatyi please review.
Signed-off-by: Aleksandr Andrienko <aandrienko@codenvy.com>

### CODE REVIEW

**Comment by AndrienkoAleksandr** [comment]
ci-build

**Comment by AndrienkoAleksandr** [comment]
ci-build

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java [hunk 1]
```diff
@@ -67,32 +67,20 @@
                                                                               service.getVolumesFrom().size());
 
             for (String dependsOn : service.getDependsOn()) {
-                if (!services.containsKey(dependsOn)) {
-                    throw new IllegalArgumentException(
-                            format("Dependency '%s' in machine '%s' points to not known machine.",
-                                   dependsOn, serviceEntry.getKey()));
-                }
+                checkDependency(dependsOn, serviceEntry.getKey(), services, "A service can not depend on itself");
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java]: We don't have 'service' entity in our public API. User would not understand what that mean if environment is not docker compose.
And message doesn't point to a machine with the error.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java [hunk 2]
```diff
@@ -67,32 +67,20 @@
...
 
             // links also counts as dependencies
             for (String link : service.getLinks()) {
                 String dependency = getServiceFromLink(link);
-                if (!services.containsKey(dependency)) {
-                    throw new IllegalArgumentException(
-                            format("Dependency '%s' in machine '%s' points to not known machine.",
-                                   dependency, serviceEntry.getKey()));
-                }
+                checkDependency(dependency, serviceEntry.getKey(), services, "A service can not link to itself");
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java]: We don't have 'service' entity in our public API. User would not understand what that mean if environment is not docker compose.
And message doesn't point to a machine with the error.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java [hunk 3]
```diff
@@ -67,32 +67,20 @@
...
             }
             // volumesFrom also counts as dependencies
             for (String volumesFrom : service.getVolumesFrom()) {
                 String dependency = getServiceFromVolumesFrom(volumesFrom);
-                if (!services.containsKey(dependency)) {
-                    throw new IllegalArgumentException(
-                            format("Dependency '%s' in machine '%s' points to not known machine.",
-                                   dependency, serviceEntry.getKey()));
-                }
+                checkDependency(dependency, serviceEntry.getKey(), services, "A service can not contain 'volumes_from' to itself");
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java]: We don't have 'service' entity in our public API. User would not understand what that mean if environment is not docker compose.
And message doesn't point to a machine with the error.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java [hunk 4]
```diff
@@ -101,48 +89,34 @@
...
-                            Optional<String> maxWeight = dependencies.get(service)
-                                                                     .stream()
-                                                                     .max((o1, o2) -> weights.get(o1).compareTo(weights.get(o2)));
-                            // optional can't be empty because size of the list is checked above
-                            //noinspection OptionalGetWithoutIsPresent
-                            weights.put(service, weights.get(maxWeight.get()) + 1);
-                            machinesLeft.remove(service);
-                            weightEvaluatedInCycleRun = true;
-                        }
+                Map.Entry<String, Set<String>> serviceEntry = it.next();
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java:96]: I think that extracting key and value here would make code more readable

**Review by garagatyi — CHANGES_REQUESTED** [review]

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java (hunk 4 above)
Comment (AndrienkoAleksandr) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/DefaultServicesStartStrategy.java:96]: Done.

**Review by garagatyi — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Comment by AndrienkoAleksandr** [comment]
ci-build

**Pull request merged**